<a href="https://colab.research.google.com/github/pastelpargoal/CodingAssessment/blob/main/Part1_HighFreq.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part 1: High Frequency Survey Checks

In [15]:
# basics
import pandas as pd
import numpy as np


In [16]:
# Reading in the file
url = "https://raw.githubusercontent.com/pastelpargoal/CodingAssessment/main/survey_results.csv"
surveydata = pd.read_csv(url)
#Checking the data contents
surveydata.head()


uniqueid     hhid  surveyid  surveyor  surveydate  surveytime  \
0         1  1101008       122  Benjamin     9242014        6308   
1         2  1101015       101     Peter     9132013        7295   
2         3  1101019       115     Peter    12142014        6339   
3         4  1101027        84  Benjamin    10172014        7430   
4         5  1101040        88      Anna     9242014        7551   

   survey_complete  superpresent  scrutinizedyn  backcheckedyn  ...  \
0                1           0.0              1            0.0  ...   
1                0           0.0              1            0.0  ...   
2                1           0.0              1            0.0  ...   
3                1           0.0              1            0.0  ...   
4                1           0.0              1            0.0  ...   

   theftvictim_2  thanavisityn     thanavisitreason  thanavisitreason_other  \
0            NaN           1.0                Other         BATWARE KE LIYE   
1            NaN           2.0                  NaN                     NaN   
2            NaN           2.0                  NaN                     NaN   
3            NaN           1.0  To register a Crime                     NaN   
4            NaN           2.0                  NaN                     NaN   

   surveytime2  thanavisitreason_other2  exit  base_age  base_income  \
0    105.13333                      NaN   1.0      44.0       1609.0   
1    121.58334                      NaN   1.0      74.0       2826.0   
2    105.65000                      NaN   1.0    1975.0       2426.0   
3          NaN                      NaN   NaN       NaN          NaN   
4    125.85000                      NaN   1.0    1983.0       1652.0   

           base_educ  
0  None or anganwadi  
1  None or anganwadi  
2                  7  
3                NaN  
4                  7  

[5 rows x 70 columns]

In [17]:
surveydata.shape

(1001, 70)

In [18]:
# Checking more details about the dataset
surveydata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 70 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   uniqueid                  1001 non-null   int64  
 1   hhid                      1001 non-null   int64  
 2   surveyid                  1001 non-null   int64  
 3   surveyor                  1001 non-null   object 
 4   surveydate                1001 non-null   int64  
 5   surveytime                1001 non-null   int64  
 6   survey_complete           1001 non-null   int64  
 7   superpresent              998 non-null    float64
 8   scrutinizedyn             1001 non-null   int64  
 9   backcheckedyn             975 non-null    float64
 10  sex                       1001 non-null   int64  
 11  age                       1000 non-null   float64
 12  educ                      989 non-null    float64
 13  income                    1001 non-null   int64  
 14  addressd

# 1. Survey Duration

## a. Calculate the average and median values of time spent surveying, considering only completed surveys.

In [19]:
#Subsetting to those that only completed the surveys
surveydata1= surveydata[surveydata.survey_complete==1]
surveydata1.head()

uniqueid     hhid  surveyid  surveyor  surveydate  surveytime  \
0         1  1101008       122  Benjamin     9242014        6308   
2         3  1101019       115     Peter    12142014        6339   
3         4  1101027        84  Benjamin    10172014        7430   
4         5  1101040        88      Anna     9242014        7551   
5         6  1101044       107      Anna    10202014        5636   

   survey_complete  superpresent  scrutinizedyn  backcheckedyn  ...  \
0                1           0.0              1            0.0  ...   
2                1           0.0              1            0.0  ...   
3                1           0.0              1            0.0  ...   
4                1           0.0              1            0.0  ...   
5                1           0.0              1            0.0  ...   

   theftvictim_2  thanavisityn     thanavisitreason  thanavisitreason_other  \
0            NaN           1.0                Other         BATWARE KE LIYE   
2            NaN           2.0                  NaN                     NaN   
3            NaN           1.0  To register a Crime                     NaN   
4            NaN           2.0                  NaN                     NaN   
5            NaN           2.0                  NaN                     NaN   

   surveytime2  thanavisitreason_other2  exit  base_age  base_income  \
0   105.133330                      NaN   1.0      44.0       1609.0   
2   105.650000                      NaN   1.0    1975.0       2426.0   
3          NaN                      NaN   NaN       NaN          NaN   
4   125.850000                      NaN   1.0    1983.0       1652.0   
5    93.933334                      NaN   1.0    1971.0       2673.0   

           base_educ  
0  None or anganwadi  
2                  7  
3                NaN  
4                  7  
5  None or anganwadi  

[5 rows x 70 columns]

In [20]:
# Avg and median
surveyAvg= float(surveydata1.surveytime.mean())
surveyMedian = surveydata1.surveytime.median()
# Displaying the results
surveyAvg


7201.995555555555

In [21]:
surveyMedian

7165.0

# 2. Enumerator Differences

## a) Does this average value mask important differences among surveyors? For each surveyor in the dataset, output the average length of time spent surveying, again for only completed surveys.

In [22]:
# Grouping the data among surveyors
totalsurveytime = surveydata1.groupby('surveyor').surveytime.sum()

In [23]:

numSurvey = len(surveydata1.surveytime)
average = totalsurveytime/ numSurvey
average

surveyor
Anna         960.632222
Benjamin    1021.962222
Caroline     454.888889
David       1138.196667
Grace        370.806667
Jane         114.264444
John         516.723333
Joseph       469.902222
Mary        1116.794444
Peter        305.738889
Sam          732.085556
Name: surveytime, dtype: float64

## b) Briefly comment on any notable differences.

### The average is 7202, which is higher than each individual surveyor. There are major differences among the surveyor with Jane having around 100 while Benjamin, David, and Mary have over 1000. Overall, there is a major difference between the average and average of each surveyor

# 3. Uniqueness of ID vars

## a. Check whether the hhid variable is unique (i.e., each observation represents a distinct household).

In [24]:
surveydata1.shape

(900, 70)

In [25]:
#Checking for uniqueness
surveydata1.hhid.unique()
surveydata1.hhid.nunique()

897

### 897 are unique while 3 are not.

## b) If duplicates exist, list the duplicated IDs.

In [26]:
surveydata1[surveydata1.hhid.duplicated()]

uniqueid     hhid  surveyid surveyor  surveydate  surveytime  \
597       598  1802011       640      Sam     9172013        7098   
641       642  1807077       981      Sam     8222014        6749   
679       680  1813023       669     Mary    10292013        6584   

     survey_complete  superpresent  scrutinizedyn  backcheckedyn  ...  \
597                1           0.0              1            1.0  ...   
641                1           0.0              1            0.0  ...   
679                1           0.0              1            0.0  ...   

     theftvictim_2  thanavisityn  thanavisitreason  thanavisitreason_other  \
597            NaN           2.0               NaN                     NaN   
641            NaN           2.0               NaN                     NaN   
679            NaN           2.0               NaN                     NaN   

     surveytime2  thanavisitreason_other2  exit  base_age  base_income  \
597    118.30000                      NaN   1.0      37.0       1508.0   
641    112.48333                      NaN   1.0    1977.0       3822.0   
679    109.73333                      NaN   1.0      37.0       2547.0   

             base_educ  
597                 12  
641  None or anganwadi  
679  None or anganwadi  

[3 rows x 70 columns]

#### Duplicates exist (3 total)

In [27]:

surveydata1.hhid.drop_duplicates()

0       1101008
2       1101019
3       1101027
4       1101040
5       1101044
         ...   
995     2024053
996     2025012
998     2025071
999     2025087
1000    2025097
Name: hhid, Length: 897, dtype: int64

## c)In a few sentences, describe what steps you might take to investigate or resolve duplicate IDs.

I would check for uniqueness and duplicates. If there is duplicates, I would investigate the dataset more (i.e., checking for shape, info, etc.) to learn more about it. Then I would drop duplicates to avoid repeated answers, to have a fair outcome.

## Note: I used my class resources to help with the coding.